[clang][driver] Suppress gnu-line-marker when saving temps

macurtis-amd opened the pull request:
When passing `-save-temps` to clang, the generated preprocessed output uses gnu line markers. This unexpectedly triggers gnu-line-marker warnings when used with `-Weverything` or `-pedantic`. Even worse, compilation fails if `-Werror` is used.

This change suppresses gnu-line-marker warnings when invoking clang with input from a preprocessor job and the user has not otherwise explictly specified `-Wgnu-line-marker` somewhere on the command line. Note that this does apply to user provided preprocessed files.

fixes #63802

macurtis-amd commented:
> I am not sure clang driver should does this unusual suppression.
> 
> You'd also get a warning with: `clang test.c -E -Wall -pedantic; clang test.i -S -Wall -pedantic`
> 
> There might be a different strategy to suppress this diagnostic?

Did you have an alternative in mind?

Driver seemed like to best place to me because it is generating the commands for both the producer and consumer of the preprocessed file.

MaskRay commented:
I don't, but I wonder what's the GCC behavior. It seems to suppress the diagnostic when processing a .i file.

macurtis-amd commented:
@MaskRay Latest revision moves the suppression out of the Driver and into the Frontend. Behavior matches gcc.

MaskRay commented on clang/test/Preprocessor/line-directive-suppressed.c:
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang_cc1 -std=c99 -fsyntax-only -pedantic %s 2>&1 | grep 'warning: this style of line directive is a GNU extension'
```
We prefer `FileCheck` to `grep`. FileCheck enables testing the line number, e.g.

```
// CHECK: [[#@LINE+1]]:4: warning: `
# 0 "zero"
```

macurtis-amd replied:
Updated test to use FileCheck.

macurtis-amd committed:
fixup! [clang] Suppress gnu-line-marker when saving temps

macurtis-amd commented:
@MaskRay Any other concerns with this change?

macurtis-amd committed:
[clang] Suppress gnu-line-marker when saving temps

When passing `-save-temps` to clang, the generated preprocessed output uses gnu
line markers. This unexpectedly triggers gnu-line-marker warnings when used with
`-Weverything` or `-pedantic`. Even worse, compilation fails if `-Werror` is
used.

This change suppresses gnu-line-marker warnings when invoking 'clang' with
preprocessor input (specified via -x argument or deduced from the input file
name). This matches gcc behavior.

fixes #63802

MaskRay reviewed (commented):
LGTM!
